src/crypto/equihash.cpp

```diff
@@ -19,183 +19,210 @@
 #include <iostream>
 #include <stdexcept>
 
-void validate_params(int n, int k)
-{
-    if (k>=n) {
-        std::cerr << "n must be larger than k\n";
-        throw invalid_params();
-    }
-    if (n % 8 != 0) {
-        std::cerr << "Parameters must satisfy n = 0 mod 8\n";
-        throw invalid_params();
-    }
-    if ((n/(k+1)) % 8 != 0) {
-        std::cerr << "Parameters must satisfy n/(k+1) = 0 mod 8\n";
-        throw invalid_params();
-    }
-}
-
-int Equihash::InitialiseState(eh_HashState& base_state)
+template<unsigned int N, unsigned int K>
+int Equihash<N,K>::InitialiseState(eh_HashState& base_state)
 {
+    unsigned int n = N;
+    unsigned int k = K;
     unsigned char personalization[crypto_generichash_blake2b_PERSONALBYTES] = {};
     memcpy(personalization, "ZcashPOW", 8);
     memcpy(personalization+8,  &n, 4);
     memcpy(personalization+12, &k, 4);
     return crypto_generichash_blake2b_init_salt_personal(&base_state,
                                                          NULL, 0, // No key.
-                                                         n/8,
+                                                         N/8,
                                                          NULL,    // No salt.
                                                          personalization);
 }
 
-StepRow::StepRow(unsigned int n, const eh_HashState& base_state, eh_index i) :
-        hash {new unsigned char[n/8]},
-        len {n/8},
-        indices {i}
+void EhIndexToArray(const eh_index i, unsigned char* array)
 {
-    eh_HashState state;
-    state = base_state;
-    crypto_generichash_blake2b_update(&state, (unsigned char*) &i, sizeof(eh_index));
-    crypto_generichash_blake2b_final(&state, hash, n/8);
+    assert(sizeof(eh_index) == 4);
+    array[0] = (i >> 24) & 0xFF;
+    array[1] = (i >> 16) & 0xFF;
+    array[2] = (i >>  8) & 0xFF;
+    array[3] =  i        & 0xFF;
```

Comment:
IIRC because then comparison was consistent with integer comparison. But I think now comparison is always done by converting first, so it actually doesn't matter how the indices are stress in the array. Unless we remove that as an optimisation, in which case it does matter.